sites/controller/__init__.py

```diff
@@ -0,0 +1,112 @@
+import bisect
+import copy
+import urllib
+
+import hca
+import sys
+import os
+import json
+import os.path
+import requests
+
+
+class FusilladeConfig:
+    def __init__(self):
+        self.api_endpoint = os.getenv('FUS_API_ENDPOINT')
+        self.swagger_endpoint = f'{self.api_endpoint}/swagger.json'
+        self.auth_client = hca.auth.AuthClient(swagger_url=self.swagger_endpoint)
+
+
+    def get_users(self):
+        return self.auth_client.get_v1_users()
+
+    def get_groups(self):
+        return self.auth_client.get_v1_groups()
+
+    def get_role(self):
+        return self.auth_client.get_v1_roles()
+
+
+#not sure if this class is needed.
+class FileController:
+    def __init__(self, file_data, file_path):
+        self.file_data = file_data
+        self.file_path = file_path
+        self.group_data = json.loads(self.file_data)
+        self.updated_data = None
+
+    def add_user_to_groups(self, user: str, groups: list):
+        """returns a new group_dictionary with user added into requested groups"""
+        modified_data = copy.deepcopy(self.group_data)
+        for group in groups:
+            try:
+                assert group in modified_data.get("groups")
+            except AssertionError:
+                print(f'group "{group}" not found in {modified_data.get("groups")} ')
+                continue
+            users_in_current_group = modified_data['groups'][group]['users']
+            if user not in users_in_current_group:
+                bisect.insort(users_in_current_group, user)
+        return modified_data
+
+class GitlabController:
+
+    def __init__(self):
+        self.access_token = os.getenv('DCP_FUS_GL_TOKEN')
+        self.gitlab_url  = os.getenv("GITLAB_URL")
+        self.access_headers = {"PRIVATE-TOKEN": self.access_token}
+        self.gitlab_project_id = os.getenv("GITLAB_PROJECT_ID")
+        self.ci_branch = self._get_ci_branch()
+
+    def get_file_from_repo(self, file_path: os.PathLike) -> requests.Response:
+        """ get a file from a repo, file_path is from root of repository"""
+        encoded_file_path = urllib.parse.quote(file_path, safe='')
+        gl_file_path = f'/api/v4/projects/{self.gitlab_project_id}/repository/files/{encoded_file_path}/raw'
+        url = f'{self.gitlab_url}{gl_file_path}'
+        parameters = {"ref": self.ci_branch}
+        resp = requests.get(url=url, headers=self.access_headers, params=parameters)
+        return resp
+
+    def create_branch(self,service_account_name):
+        gl_branch_path = f'/projects/{self.gitlab_project_id}/repository/branches'
+        parameters = {"branch": f'{self.ci_branch}-{service_account_name}',
```

Comment:
Suggest making these branches timestamped, in case the operator attempts to modify the same service account multiple times.

```suggestion
        branch_name = f"{self.ci_branch}-{service_account_name}-{datetime.today().strftime('%Y%m%d-%H%M%S')}"
        parameters = {"branch": f'{self.ci_branch}-{service_account_name}',
```

Reply:
Should probably split this off into a `_get_branch_name()` and call it in the constructor, so the `commit_changes()` method has access to the branch name too.

Reply:
One more question... can git handle `@` symbols in a branch name? The `@` symbol is reserved in git (has a special meaning), so I think you'd need to replace `@`s with `_`s or something similar.

Reply:
Alternatively, just use the timestamp suggestion to make sure this branch name is unique, and don't include the service account email in the branch name.